metadata/src/main/java/org/apache/kafka/controller/ActivationRecordsGenerator.java

```diff
@@ -0,0 +1,197 @@
+            switch (featureControl.zkMigrationState()) {
+                case NONE:
+                    // Since this is the default state there may or may not be an actual NONE in the log. Regardless,
+                    // it will eventually be persisted in a snapshot, so we don't need to explicitly write it here.
+                    if (zkMigrationEnabled) {
+                        throw new RuntimeException("Should not have ZK migrations enabled on a cluster that was " +
+                            "created in KRaft mode.");
+                    }
+                    break;
+                case PRE_MIGRATION:
+                    log.warn("Activating pre-migration controller without empty log. There may be a partial migration");
+                    break;
+                case MIGRATION:
+                    if (!zkMigrationEnabled) {
+                        // This can happen if controller leadership transfers to a controller with migrations enabled
+                        // after another controller had finalized the migration. For example, during a rolling restart
+                        // of the controller quorum during which the migration config is being set to false.
+                        log.warn("Completing the ZK migration since this controller was configured with " +
+                            "'zookeeper.metadata.migration.enable' set to 'false'.");
+                        records.add(ZkMigrationState.POST_MIGRATION.toRecord());
+                    } else {
+                        log.info("Staying in the ZK migration since 'zookeeper.metadata.migration.enable' is still " +
+                            "'true'.");
+                    }
+                    break;
+                case POST_MIGRATION:
+                    if (zkMigrationEnabled) {
+                        log.info("Ignoring 'zookeeper.metadata.migration.enable' value of 'true' since the ZK " +
+                            "migration has been completed.");
+                    }
+                    break;
+                default:
+                    throw new IllegalStateException("Unsupported ZkMigrationState " + featureControl.zkMigrationState());
+            }
+        } else {
+            if (zkMigrationEnabled) {
+                throw new RuntimeException("Should not have ZK migrations enabled on a cluster running " +
+                    "metadata.version " + featureControl.metadataVersion());
+            }
+        }
+
+        if (transactionStartOffset != -1L) {
+            if (!metadataVersion.isMetadataTransactionSupported()) {
+                throw new RuntimeException("Detected in-progress transaction at " + transactionStartOffset +
+                    ", but the metadata.version " + metadataVersion +
+                    " does not support transactions. Cannot continue.");
+            } else {
+                log.warn("Detected in-progress transaction at " + transactionStartOffset +
```

Comment:
maybe would be good to specify "in-progress metadata transaction" to avoid confusion with EOS transactions